### PR TITLE
Update rebalance tokenIds logic

### PR DESCRIPTION
**Pull request by prat-gpt**
# Description of bug or vulnerability and solution
* In case of partial settle, `borrowerTokenIds.length * 1e18 < borrower.collateral`, this leads to failing `invariant_CT4`.
* Solved by rebalancing tokenIds only when collateral is fully used.

# Contract size
## Pre Change
ERC721Pool               -  24,492B  (99.65%)
## Post Change
ERC721Pool               -  24,516B  (99.75%)

### CODE REVIEW

File: src/ERC721Pool.sol [hunk 1]
```diff
@@ -399,7 +399,9 @@ contract ERC721Pool is FlashloanablePool, IERC721Pool {
             params
         );
 
-        if (result.collateralSettled > 0) _rebalanceTokens(params.borrower, result.collateralRemaining);
+        // move balances from borrower token ids array to pool / claimable token ids array
+        // if entire bad debt was settled or if all available collateral was used to settle bad debt
+        if(result.collateralSettled != 0) _rebalanceTokens(params.borrower, result.collateralRemaining);
```
Comment (grandizzy) [inline src/ERC721Pool.sol]: nit: space needed after `if`

File: src/ERC721Pool.sol [hunk 2]
```diff
@@ -399,7 +399,9 @@ contract ERC721Pool is FlashloanablePool, IERC721Pool {
             params
         );
 
-        if (result.collateralSettled > 0) _rebalanceTokens(params.borrower, result.collateralRemaining);
+        // move balances from borrower token ids array to pool / claimable token ids array
```
Comment (grandizzy) [inline src/ERC721Pool.sol]: this comment is outdated, I suggest
```
move token ids from borrower array to pool claimable array if any collateral used to settle bad debt
```

**Review by grandizzy — COMMENTED** [review]
Lgtm, left a comment

File: src/ERC721Pool.sol (hunk 1 above)
Comment (prat-gpt) [reply src/ERC721Pool.sol]: done here [03247ae](https://github.com/ajna-finance/contracts/pull/746/commits/03247ae260fac4d7d1e8ad90c61daf01330ae4cc)

File: src/ERC721Pool.sol (hunk 2 above)
Comment (prat-gpt) [reply src/ERC721Pool.sol]: done here [03247ae](https://github.com/ajna-finance/contracts/pull/746/commits/03247ae260fac4d7d1e8ad90c61daf01330ae4cc)

**Review by grandizzy — APPROVED** [review]
Lgtm

File: src/ERC721Pool.sol [hunk 3]
```diff
@@ -576,7 +577,11 @@ contract ERC721Pool is FlashloanablePool, IERC721Pool {
         uint256[] storage borrowerTokens = borrowerTokenIds[borrowerAddress_];
 
         uint256 noOfTokensPledged    = borrowerTokens.length;
-        uint256 noOfTokensToTransfer = borrowerCollateral_ != 0 ? noOfTokensPledged - borrowerCollateral_ / 1e18 : noOfTokensPledged;
+        /*
+            eg1. borrowerCollateral_ = 4.1, noOfTokensPledged = 6; noOfTokensToTransfer = 1
+            eg2. borrowerCollateral_ = 4, noOfTokensPledged = 6; noOfTokensToTransfer = 2
+        */
+        uint256 noOfTokensToTransfer = borrowerCollateral_ != 0 ? noOfTokensPledged - (borrowerCollateral_ + 1e18 - 1) / 1e18 : noOfTokensPledged;
```
Comment (mattcushman) [inline src/ERC721Pool.sol]: For legibility, what do you think of doing:
```
uint256 borrowerCollateralRoundedUp = (borrowerCollateral_ + 1e18 - 1) / 1e18;
uint256 noOfTokensToTransfer = noOfTokensPledged - borrowerColleralRoundedUp;
```

**Review by mattcushman — APPROVED** [review]
LGTM.  made one suggestion to make the code more legible.

**Pull request merged**